siatest/testgroup.go

```diff
@@ -0,0 +1,341 @@
+package siatest
+
+import (
+	"errors"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/node"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/fastrand"
+)
+
+type (
+	// GroupParams is a helper struct to make creating TestGroups easier.
+	GroupParams struct {
+		hosts   int // number of hosts to create
+		renters int // number of renters to create
+		miners  int // number of miners to create
+	}
+
+	// TestGroup is a group of of TestNodes that are funded, synced and ready
+	// for upload, download and mining depending on their configuration
+	TestGroup struct {
+		nodes   map[*TestNode]struct{}
+		hosts   map[*TestNode]struct{}
+		renters map[*TestNode]struct{}
+		miners  map[*TestNode]struct{}
+
+		dir string
+	}
+)
+
+var (
+	// defaultAllowance is the allowance used for the group's renters
+	defaultAllowance = modules.Allowance{
+		Funds:       types.SiacoinPrecision.Mul64(100000),
+		Hosts:       10,
+		Period:      20,
+		RenewWindow: 5,
+	}
+)
+
+// announceHosts adds storage to each host and announces them to the group
+func announceHosts(hosts map[*TestNode]struct{}) error {
+	for host := range hosts {
+		if err := host.HostStorageFoldersAddPost(host.Dir, 1048576); err != nil {
+			return err
+		}
+		if err := host.HostAcceptingContractsPost(true); err != nil {
+			return err
+		}
+		if err := host.HostAnnouncePost(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Close closes the group and all its nodes
+func (tg *TestGroup) Close() (err error) {
+	for n := range tg.nodes {
+		err = build.ComposeErrors(err, n.Close())
+	}
+	return
+}
+
+// fullyConnectNodes takes a list of nodes and connects all their gateways
+func fullyConnectNodes(nodes []*TestNode) error {
+	// Fully connect the nodes
+	for i, nodeA := range nodes {
+		for _, nodeB := range nodes[i+1:] {
+			if err := nodeA.GatewayConnectPost(nodeB.GatewayAddress()); err != nil {
+				return build.ExtendErr("failed to connect to peer", err)
```

Comment:
Is it possible that this returns 'already a peer' errors?

Reply:
Hm... that could probably happen if you have the nodes A, B and C and they connect in the following order: AB, AC, BC. 
B and C might already have connected due to being connect via A. This is probably not very likely for small groups but I will add a check.